libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2213,7 +2213,7 @@ void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCa
 				externalFunctionType->declaration().scope() == m_currentContract
```

Comment:
Can you check if `scope()` returns `nullptr` for free functions? If it does, we also need a check against `m_currentContract` being null here. When it's a free function, the compiler should not suggest adding `this.`

I think this part does not show up in syntax tests so please also test this by running compiled `solc` locally on one of the test cases. You should see the secondary message that way.

Reply:
I did a `dynamic_cast` check and the `scope()` of a free function is not `nullptr`. It is the `SourceUnit`. Also, from my understanding, free functions would not pass the `visibility()` check above because free functions are not `public`.

Reply:
True. They'll fail the first check. Good.